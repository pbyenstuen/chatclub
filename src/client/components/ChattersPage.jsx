import React from 'react';
import useLoader from "./useLoader";
import ChattersList from './ChattersList';
import CreateChatterForm from './CreateChatterForm';

const ChattersPage = ({ api }) => {
    const { data: chatters, loading, error, reload: updateList } = useLoader(async () => await api.chatters.getChatters());

    return (
        <>
            <header>
                <h2>Chatters</h2>
            </header>
            <div id="chatters-page">
                <CreateChatterForm
                    api={api}
                    updateList={updateList}
                />
                <ChattersList
                    api={api}
                    chatters={chatters}
                    loading={loading}
                    error={error}
                    updateList={updateList}
                />
            </div>
        </>
    )
}

export default ChattersPage;
