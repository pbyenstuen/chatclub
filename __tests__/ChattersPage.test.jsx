import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import ChattersPage from "../src/client/components/ChattersPage";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("ChattersPage", () => {
    it("it shows created chatters", async () => {
        const api = {
            chatters: {
                getChatters: () => [{
                    id: 1,
                    email: "test@test.test",
                    firstName: "Test",
                    lastName: "Tester"
                }]
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("p").textContent).toEqual("test@test.test");
    });

    it("can create new chatter", async () => {
        const createChatter = jest.fn();

        const api = {
            chatters: {
                createChatter,
                getChatters: () => [{
                    id: 1,
                    email: "test@test.test",
                    firstName: "Test",
                    lastName: "Tester"
                }]
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        Simulate.change(container.querySelector("input[name='email']"), {
            target: { value: "chat@chat.chat" },
        });

        Simulate.change(container.querySelector("input[name='firstname']"), {
            target: { value: "Chat" },
        });

        Simulate.change(container.querySelector("input[name='lastname']"), {
            target: { value: "Chatson" },
        });

        await act(async () => {
            Simulate.submit(container.querySelector("form"));
        });

        expect(createChatter).toBeCalledWith({
            email: "chat@chat.chat",
            firstName: "Chat",
            lastName: "Chatson"
        });
    });

    it("can show error message for empty fields when creating chatter", async () => {
        const api = {
            chatters: {
                getChatters: () => [{
                    id: 1,
                    email: "test@test.test",
                    firstName: "Test",
                    lastName: "Tester"
                }]
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        Simulate.submit(container.querySelector("form"));

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h5").textContent).toEqual("Please enter all fields");
    });

    it("can show error message on creating chatter", async () => {
        const api = {
            chatters: {
                getChatters: () => [{
                    id: 1,
                    email: "test@test.test",
                    firstName: "Test",
                    lastName: "Tester"
                }],
                createChatter: () => {
                    throw new Error("Error");
                }
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        container.querySelectorAll("input").forEach(e => 
            Simulate.change(e, {
            target: { value: "Value" },
        }));

        await act(async () => {
            Simulate.submit(container.querySelector("form"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h5").textContent).toEqual("Error: Error");
    });

    it("can show error message on rendering chatters list", async () => {
        const api = {
            chatters: {
                getChatters: () => {
                    throw new Error("Error");
                }
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("Something went wrong: Error: Error");
    });

    it("can show edit chatter page", async () => {
        const api = {
            chatters: {
                getChatters: () => [{
                    id: 1,
                    email: "test@test.test",
                    firstName: "Test",
                    lastName: "Tester"
                }]
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        await act(async () => {
            await Simulate.click(container.querySelector("#chatter-item a"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("Edit Chatter");
        expect(container.querySelector("h3").textContent).toEqual("test@test.test");
    });

    it("can call deletion function", async () => {
        const deleteChatter = jest.fn();

        const api = {
            chatters: {
                deleteChatter,
                getChatters: () => [{
                    id: 1,
                    email: "test@test.test",
                    firstName: "Test",
                    lastName: "Tester"
                }]
            }
        }

        await mountComponent(<ChattersPage api={api} />);

        await act(async () => {
            await Simulate.click(container.querySelector("#del-btn"));
        });

        expect(deleteChatter).toBeCalled();
    });
});
