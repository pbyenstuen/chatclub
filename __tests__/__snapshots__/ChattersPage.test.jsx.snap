// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChattersPage can show error message for empty fields when creating chatter 1`] = `"<header><h2>Chatters</h2></header><div id=\\"chatters-page\\"><div id=\\"create-chatter-form\\"><h4>CREATE NEW CHATTER</h4><form><h5>Please enter all fields</h5><div><label>Email <input type=\\"email\\" name=\\"email\\" value=\\"\\"></label></div><div><label>First name <input type=\\"text\\" name=\\"firstname\\" value=\\"\\"></label></div><div><label>Last name <input type=\\"text\\" name=\\"lastname\\" value=\\"\\"></label></div><button class=\\"submit-btn\\" type=\\"submit\\">CREATE CHATTER</button></form></div><div id=\\"chatters-container\\"><div id=\\"chatter-item\\"><h5>Test Tester</h5><p>test@test.test</p><button id=\\"edit-btn\\"><a href=\\"/chatters/1/edit\\">EDIT</a></button><button id=\\"del-btn\\">DELETE</button></div></div></div>"`;

exports[`ChattersPage can show error message on creating chatter 1`] = `"<header><h2>Chatters</h2></header><div id=\\"chatters-page\\"><div id=\\"create-chatter-form\\"><h4>CREATE NEW CHATTER</h4><form><h5>Error: Error</h5><div><label>Email <input type=\\"email\\" name=\\"email\\" value=\\"Value\\"></label></div><div><label>First name <input type=\\"text\\" name=\\"firstname\\" value=\\"Value\\"></label></div><div><label>Last name <input type=\\"text\\" name=\\"lastname\\" value=\\"Value\\"></label></div><button class=\\"submit-btn\\" type=\\"submit\\">CREATE CHATTER</button></form></div><div id=\\"chatters-container\\"><div id=\\"chatter-item\\"><h5>Test Tester</h5><p>test@test.test</p><button id=\\"edit-btn\\"><a href=\\"/chatters/1/edit\\">EDIT</a></button><button id=\\"del-btn\\">DELETE</button></div></div></div>"`;

exports[`ChattersPage it shows created chatters 1`] = `"<header><h2>Chatters</h2></header><div id=\\"chatters-page\\"><div id=\\"create-chatter-form\\"><h4>CREATE NEW CHATTER</h4><form><div><label>Email <input type=\\"email\\" name=\\"email\\" value=\\"\\"></label></div><div><label>First name <input type=\\"text\\" name=\\"firstname\\" value=\\"\\"></label></div><div><label>Last name <input type=\\"text\\" name=\\"lastname\\" value=\\"\\"></label></div><button class=\\"submit-btn\\" type=\\"submit\\">CREATE CHATTER</button></form></div><div id=\\"chatters-container\\"><div id=\\"chatter-item\\"><h5>Test Tester</h5><p>test@test.test</p><button id=\\"edit-btn\\"><a href=\\"/chatters/1/edit\\">EDIT</a></button><button id=\\"del-btn\\">DELETE</button></div></div></div>"`;
